### DISCUSSION TITLE
Intergrating a user-designed linear solver

### DISCUSSION THREAD

**Discussion by rebpoli** [libMesh/libmesh#3036] [Q&A] [answered]
Hi all,

I have a linear solver I'd like to test in libmesh.
Can you provide some directions on how to integrate it?
What are the main hooks I should look at?
I am using linear and nonlinear implicit systems.

Thanks,
Renato

**Comment by jwpeterson** [comment]
I guess you are referring to implementing a "meta" solver that would combine some existing PETSc (or other) linear solvers in a new way, and not implementing your own linear solver from scratch? The `LinearSolver` class hierarchy in libmesh has specific implementations for PETSc, Eigen, and others, so I would probably start there.

**Reply by rebpoli** [reply]
It is a separate solver. I will take a look at LinearSolver, thanks.

**Comment by roystgnr** [comment, marked as answer]
Agreed with John that subclassing LinearSolver may be the best way to start for you.  After you have things working for linear problems, you could try using your LinearSolver within a libMesh::NewtonSolver and attaching that to your NonlinearImplicitSystem, then giving your custom LinearSolver to it.  That'll at least see how things perform with your linear solver wrapped in a quasi-Newton + line search outer solver.  I'm not sure how best to connect a custom linear solver to the more sophisticated nonlinear solvers in PETSc SNES, though.

**Reply by rebpoli** [reply]
I will dig into that. I understand your point.
Thanks.
